crates/namada/src/ledger/native_vp/masp.rs

```diff
@@ -304,110 +296,327 @@ where
+        None if changed_balances.tokens.contains_key(&out.asset_type) => {
+            // Otherwise note the contribution to this transparent output
+            let (token, _denom, digit) =
+                &changed_balances.tokens[&out.asset_type];
+            let amount =
+                token::Amount::from_masp_denominated(out.value, *digit);
+            *bal_ref = bal_ref
+                .checked_sub(&ValueSum::from_pair(token.clone(), amount))
+                .ok_or_else(|| {
+                    Error::NativeVpError(native_vp::Error::SimpleMessage(
+                        "Overflow in bundle balance",
+                    ))
+                })?;
+        }
+        // unrecognized asset
+        _ => {
+            let error = Error::NativeVpError(native_vp::Error::SimpleMessage(
+                "Unable to decode asset type",
+            ));
+            tracing::debug!("{error}");
+            return Err(error);
+        }
+    };
+    Ok(())
+}
+
+// Update the transaction value pool and also ensure that the Transaction is
+// consistent with the balance changes. I.e. the transparent inputs are not more
+// than the initial balances and that the transparent outputs are not more than
+// the final balances.
+fn validate_transparent_bundle(
+    shielded_tx: &Transaction,
+    changed_balances: &mut ChangedBalances,
+    transparent_tx_pool: &mut I128Sum,
+    epoch: Epoch,
+    conversion_state: &ConversionState,
+    signers: &mut BTreeSet<TransparentAddress>,
+) -> Result<()> {
+    if let Some(transp_bundle) = shielded_tx.transparent_bundle() {
+        for vin in transp_bundle.vin.iter() {
+            validate_transparent_input(
+                vin,
+                changed_balances,
+                transparent_tx_pool,
+                epoch,
+                conversion_state,
+                signers,
+            )?;
+        }
+
+        for out in transp_bundle.vout.iter() {
+            validate_transparent_output(
+                out,
+                changed_balances,
+                transparent_tx_pool,
+                epoch,
+                conversion_state,
+            )?;
+        }
+    }
+    Ok(())
+}
+
+// Verify that the pre balance + the Sapling value balance = the post balance
+// using the decodings in tokens and conversion_state for assistance.
+fn verify_sapling_balancing_value(
+    pre: &ValueSum<Address, Amount>,
+    post: &ValueSum<Address, Amount>,
+    sapling_value_balance: &I128Sum,
+    target_epoch: Epoch,
+    tokens: &BTreeMap<AssetType, (Address, token::Denomination, MaspDigitPos)>,
+    conversion_state: &ConversionState,
+) -> Result<()> {
+    let mut acc = ValueSum::<Address, SignedAmount>::from_sum(pre.clone());
+    for (asset_type, val) in sapling_value_balance.components() {
+        // Only assets with at most the target timestamp count
+        match conversion_state.assets.get(asset_type) {
+            Some(((address, _, digit), asset_epoch, _, _))
```

Comment:
same comment

Reply:
Resolved in https://github.com/anoma/namada/pull/3186#discussion_r1606996403 .